From 4873746fa9d42a8edbc1e192899e00c29ed3d32a Mon Sep 17 00:00:00 2001
From: Giulio Benetti <giulio.benetti@benettiengineering.com>
Date: Mon, 17 Jan 2022 12:26:00 +0100
Subject: [PATCH] main.c: fix building with Linux version >= 5.13

of_get_mac_address() on Linux version >= 5.13 requires mac pointer as
second argument and return an int. So let's deal with it by checking linux
version to make it compatible with both Linux version >= 5.13 and not.

Signed-off-by: Giulio Benetti <giulio.benetti@benettiengineering.com>
---
 main.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/main.c b/main.c
index b60e18d..06cb1f6 100644
--- a/drivers/net/wireless/xradio/main.c
+++ b/drivers/net/wireless/xradio/main.c
@@ -14,6 +14,7 @@
 #include <net/cfg80211.h>
 #include <linux/of_net.h>
 #include <linux/mmc/sdio_func.h>
+#include <linux/version.h>
 
 #include "xradio.h"
 #include "fwio.h"
@@ -501,8 +502,12 @@ int xradio_core_init(struct sdio_func* f
 	u8 b;		/* MRK 5.5a */
 	struct ieee80211_hw *dev;
 	struct xradio_common *hw_priv;
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 13, 0)
+	unsigned char addr[ETH_ALEN];
+#else
 	unsigned char randomaddr[ETH_ALEN];
 	const unsigned char *addr = NULL;
+#endif
 
 	//init xradio_common
 	dev = xradio_init_common(sizeof(struct xradio_common));
@@ -515,6 +520,16 @@ int xradio_core_init(struct sdio_func* f
 	hw_priv->sdio_func = func;
 	sdio_set_drvdata(func, hw_priv);
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 13, 0)
+	// fill in mac addresses
+	if (hw_priv->pdev->of_node) {
+		err = of_get_mac_address(hw_priv->pdev->of_node, addr);
+	}
+	if (err < 0) {
+		dev_warn(hw_priv->pdev, "no mac address provided, using random\n");
+		eth_random_addr(addr);
+	}
+#else
 	// fill in mac addresses
 	if (hw_priv->pdev->of_node) {
 		addr = of_get_mac_address(hw_priv->pdev->of_node);
@@ -524,6 +539,7 @@ int xradio_core_init(struct sdio_func* f
 		eth_random_addr(randomaddr);
 		addr = randomaddr;
 	}
+#endif
 	for (b = 0; b < XRWL_MAX_VIFS; b++) {				/* MRK 5.5a */
 		memcpy(hw_priv->addresses[b].addr, addr, ETH_ALEN);
 		hw_priv->addresses[b].addr[5] += b;
